[MRTK Shell Parity] [Button] Fit & finish: proximity light intensity, distance, pulse effect polish

cre8ivepark opened the pull request:
## Overview
Fit & finish: proximity light intensity, distance, pulse effect polish
Part of #4200 [MRTK Shell Parity] UX Controls Parity with HoloLens 2 Shell  

## Changes
- Fixes: #4410 

## Updates to match shell's design spec
- Updated front cage's borderline thickness, glow intensity (PressableButton prefab)
- Icon and Text moves the 1/2 distance of the front cage (PressableButton.cs)
- Updated example button sizes (HandInteractionExample scene)
- Updated Toggle Features panel and added it as a prefab

![MRTK_Button_Polish_IconText2](https://user-images.githubusercontent.com/13754172/58124689-a5ae1800-7bc3-11e9-865b-9e9142a40168.gif)

![MRTK_Button_Polish_Lighting](https://user-images.githubusercontent.com/13754172/58124802-ec9c0d80-7bc3-11e9-8c9c-f9669c07f16a.gif)

![2019-05-21 12_04_05-Unity 2018 4 0f1 Personal - HandInteractionExamples unity - MRTK-Public-Microsof](https://user-images.githubusercontent.com/13754172/58124850-063d5500-7bc4-11e9-8832-0eb7ffe97675.png)

julenka commented:
Why wouldn't we make pressablebutton have the nice background the pressablebutton plated has? Right now pressablebutton is transparent. I imagine most people will think they need to add pressablebutton, and not pressablebutton plated.

julenka commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/PressableButtons/PressableButton.cs:
```diff
@@ -23,6 +23,10 @@ public class PressableButton : MonoBehaviour, IMixedRealityTouchHandler
         [Tooltip("The object that is being pushed.")]
         private GameObject movingButtonVisuals = null;
 
+        [SerializeField]
```
I can see how you needed this to implement the specific behavior you needed for the shell buttons, but is it something you think should be in the generic pressable button, the thing that most developers use?

It seems like the specific behavior you want is to position the object exactly half way along the press amount. For this very specific behavior it may make more sense to have a component on the moving icon that positions itself based on the press direction of the button and the press amount.

cre8ivepark replied:
Since the intention of this control is to provide the shell-parity, great default experience, we would like to include it here so that developers can get it free when it is used without modification. From the customization standpoint, I agree that we can expose this as a configurable element.

cre8ivepark replied:
Forgot to mention that this fixes the bug of icon & text being occluded by the back plate when pressed fully. Current button has a wrong thickness dimension which made the icon & text still visible on a full press.

julenka replied:
Hm, I actually didn't realize that the intention of PressableButton.cs was to provide shell parity button. I thought PressableButton.cs intention was to provide a pressable button, meaning a button that has some travel amount when touched, and then fires press and release events based on thresholds. I think that might be where my  comments are coming from. I wonder if the only PressableButton behavior we provide should require to match the shell so strictly in behavior. Instead it could be better to make a  ShellParityButton.cs that extends PressableButton which would have this behavior.


For example, you could do something like the following:


```
public class HoloLens2ShellParityButton : PressableButton
{
    [SerializeField] MovingButtonIconText = null;

    protected void UpdateMovingVisualsPosition()
    {
        base.UpdateMovingVisualsPosition();
        if (movingButtonIconText != null)
        {
            Debug.Assert(initialTransform != null);
            movingButtonIconText.transform.position = initialTransform.position + WorldSpacePressDirection * currentPushDistance / 2;
        }
    }
}
```

cre8ivepark replied:
Now I understand. Essentially, we would like to provide a baseline PressableButton which developers can leverage it for implementing any custom buttons. PressableButtonHoloLens2 and PressableButtonRound are great examples. Might be better to rename prefabs to clarify this.

![2019-05-23 14_54_06-Unity 2018 3 8f1 Personal - HandInteractionExamples unity - MRTK-Public-Microsof](https://user-images.githubusercontent.com/13754172/58289335-ec8a4200-7d6a-11e9-8b6c-f799b0e74add.png)

cre8ivepark replied:
Applied the updates! (description on the main comment thread)

julenka commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/PressableButtons/PressableButton.cs:
```diff
@@ -23,6 +23,10 @@ public class PressableButton : MonoBehaviour, IMixedRealityTouchHandler
 
+        [SerializeField]
+        [Tooltip("The icon and text content moving inside the button.")]
+        private GameObject movingButtonIconText = null;
```
If you think we should put this in the general pressable button, then make sure to expose this so that developers can configure this at runtime.

cre8ivepark replied:
Sounds great!

cre8ivepark commented:
Having PressableButton without the plate is for the scenarios where we have multiple buttons on the shared plate, leveraging the nested prefab. Might be better to have the name PressableButton as plated default and PressableButtonWithoutPlate.

![2019-05-22 17_51_42-Unity 2018 3 8f1 Personal - HandInteractionExamples unity - MRTK-Public-Microsof](https://user-images.githubusercontent.com/13754172/58218311-5814d680-7cbb-11e9-98a6-31f5ef9f9d5c.png)

cre8ivepark commented:
I tried renaming current PressableButton.prefab as PressableButtonNoPlate and PressableButtonPlated as PressableButton. However, git recognizes one of these changes as deleting & creating a new file instead of renaming. Not sure what's the best way to rename them without breaking existing users.

Cameron-Micka commented:
> I tried renaming current PressableButton.prefab as PressableButtonNoPlate and PressableButtonPlated as PressableButton. However, git recognizes one of these changes as deleting & creating a new file instead of renaming. Not sure what's the best way to rename them without breaking existing users.

Hi Yoon, I'm going to pull these changes down now and give them a try. Thanks for doing this! The rename should be fine as long as the GUID stays the same. So, if you do a rename in the Unity editor the GUID should remain the same (you can inspect the associated meta file to verify).

Cameron-Micka commented on Assets/MixedRealityToolkit.SDK/Features/UX/Interactable/Prefabs/PressableButton.prefab:
```diff
@@ -1400,6 +1400,38 @@ MeshRenderer:
   m_SortingLayerID: 0
   m_SortingLayer: 0
   m_SortingOrder: 0
+--- !u!1 &8352633413104042190
```
One question that has come up with partners and the pressable button is what is the purpose of the "Static Content?" Is that something we can delete while we are working in here? It seems to always be disabled.

![image](https://user-images.githubusercontent.com/13305729/58264363-97cad500-7d32-11e9-9636-af1588363a32.png)

cre8ivepark replied:
We can delete them. Jonathan placed them as a placeholder when the design was not finalized. I'll update them.

Cameron-Micka replied:
Awesome! Thank you Yoon.

cre8ivepark replied:
Removed redundant elements. Updated plate names for easier understanding.
![2019-05-23 15_45_30-Unity 2018 3 8f1 Personal - HandInteractionExamples unity - MRTK-Public-Microsof](https://user-images.githubusercontent.com/13754172/58291981-6f63ca80-7d74-11e9-9144-902b740f175c.png)

cre8ivepark commented:
Updates per review feedback:
- Reverted PressableButton.cs and Added new script **PressableButtonHoloLens2**.cs which contains shell design specific elements (icon & text movments) 
- Renamed PressableButtonPlated.prefab to **PressableButtonHoloLens2.prefab**
- Renamed PressableButton.prefab to **PressableButtonHoloLens2Unplated.prefab**
- Updated AppBar prefab which uses PressableButton prefab.
- No GUID changes. Developer should be able to update without breaking anything.

cre8ivepark commented:
@Cameron-Micka Unchecked both HoverLight and ProximityLight on the backplate.

cre8ivepark commented:
I'll test these updates with the merge with Bernadette's PR #4527

julenka commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/PressableButtons/PressableButton.cs:
```diff
@@ -60,7 +60,7 @@ public class PressableButton : MonoBehaviour, IMixedRealityTouchHandler
         private const float MaxRetractDistanceBeforeReset = 0.0001f;
 
-        private float currentPushDistance = 0.0f;
+        protected float currentPushDistance = 0.0f;
```
This actually seems like something that would be useful to expose as public property (getter only) rather than make protected. For example for a script that wants to animate something based on pressed amount.

cre8ivepark replied:
Made it public. Thanks!

cre8ivepark commented:
@julenka I have been noticing the same issue when I keep working on a scene and run play mode multiple times - framerate drops and the camera movement becomes slow. Not sure if it is MRTK issue or Unity issue. It runs smoothly when you restart. (even with all objects are in a viewport)
I'll also check on the device.

![2019-05-24 11_48_26-Unity 2018 3 8f1 Personal - HandInteractionExamples unity - MRTK-Public-Microsof](https://user-images.githubusercontent.com/13754172/58350351-7b09cc80-7e1a-11e9-9f6c-7251b5f2b133.png)

cre8ivepark commented:
Confirmed it runs @ 60fps on the device.

cre8ivepark commented:
## Summary
- Added new script **PressableButtonHoloLens2.cs** which contains shell design specific elements (icon & text movments). (inherits PressableButton.cs)
- Renamed PressableButtonPlated.prefab to **PressableButtonHoloLens2**.prefab
- Renamed PressableButton.prefab to **PressableButtonHoloLens2Unplated**.prefab
- PressableButtonHoloLens2 and PressableButtonHoloLens2Unplated now use **PressableButtonHoloLens2** script.
- Updated AppBar prefab which uses PressableButton prefab.
- **No GUID changes**. Developer should be able to update without breaking anything.
- Existing Piano keys and Round buttons are still using PressableButton.cs script which are good example of customization.

thalbern commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/PressableButtons/PressableButton.cs:
```diff
@@ -94,7 +94,8 @@ public SpaceMode DistanceSpaceMode
         private const float MaxRetractDistanceBeforeReset = 0.0001f;
 
-        private float currentPushDistance = 0.0f;
+        public float currentPushDistance = 0.0f;
```
this can be private

cre8ivepark replied:
Made it public per suggestion - it could be useful for the developers who want to leverage current push distance.

julenka commented:
@Cameron-Micka could you please take a look to see if your blocking issues have been addressed?

cre8ivepark commented:
Finished updating the Button README file as well to reflect naming updates.

wiwei commented:
Merging despite @Cameron-Micka's review, will followup afterward.
